x/withdraw/keeper/ibc_callbacks.go

```diff
@@ -0,0 +1,209 @@
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	vestexported "github.com/cosmos/cosmos-sdk/x/auth/vesting/exported"
+
+	transfertypes "github.com/cosmos/ibc-go/v3/modules/apps/transfer/types"
+	clienttypes "github.com/cosmos/ibc-go/v3/modules/core/02-client/types"
+	channeltypes "github.com/cosmos/ibc-go/v3/modules/core/04-channel/types"
+	"github.com/cosmos/ibc-go/v3/modules/core/exported"
+
+	evmos "github.com/tharsis/evmos/v2/types"
+	"github.com/tharsis/evmos/v2/x/withdraw/types"
+)
+
+// OnRecvPacket performs an IBC receive callback. It returns the tokens that
+// users transferred to their Cosmos secp256k1 address instead of the Ethereum
+// ethsecp256k1 address. The expected behavior is as follows:
+//
+// First transfer from authorized source chain:
+// - sends back IBC tokens which originated from the source chain
+// - sends over all Evmos native tokens
+// Second transfer from a different authorized source chain:
+// - only sends back IBC tokens which originated from the source chain
+func (k Keeper) OnRecvPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	ack exported.Acknowledgement,
+) exported.Acknowledgement {
+	logger := k.Logger(ctx)
+
+	params := k.GetParams(ctx)
+	claimsParams := k.claimsKeeper.GetParams(ctx)
+
+	// check channels from this chain (i.e destination)
+	if !params.EnableWithdraw ||
+		!claimsParams.IsAuthorizedChannel(packet.DestinationChannel) ||
+		claimsParams.IsEVMChannel(packet.DestinationChannel) {
+		// return original ACK if:
+		// - withdraw is disabled globally
+		// - channel is not authorized
+		// - channel is an EVM channel
+		return ack
+	}
+
+	// unmarshal packet data to obtain the sender and recipient
+	var data transfertypes.FungibleTokenPacketData
+	if err := transfertypes.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		err = sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "cannot unmarshal ICS-20 transfer packet data")
+		return channeltypes.NewErrorAcknowledgement(err.Error())
+	}
+
+	// validate the sender bech32 address from the counterparty chain
+	bech32Prefix := strings.Split(data.Sender, "1")[0]
+	if bech32Prefix == data.Sender {
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid sender: %s", data.Sender).Error(),
+		)
+	}
+
+	senderBz, err := sdk.GetFromBech32(data.Sender, bech32Prefix)
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid sender %s, %s", data.Sender, err.Error()).Error(),
+		)
+	}
+
+	// change the bech32 human readable prefix (HRP) of the sender to `evmos1`
+	sender := sdk.AccAddress(senderBz)
+
+	// obtain the evmos recipient address
+	recipient, err := sdk.AccAddressFromBech32(data.Receiver)
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid receiver address %s", err.Error()).Error(),
+		)
+	}
```

Comment:
bunch of shared code from `x/claims` - can we somehow reuse between the two modules without maintaining two implementation? I'm worried that with duplicate validation code and having to maintain two pieces that should be doing the same thing, we may miss updating both of these places one day.

Reply:
refactored 👍